chore; Update README with comprehensive documentation and badges

kiwamizamurai opened the pull request:
First of all, thank you for such a great product.

In some versions of NextAuth, there have been changes to certain parameters such as session_name, which caused me some difficulties in adapting to the library's specifications. I have also confirmed that others have faced similar challenges through related issues on the repository. https://github.com/TCatshoek/fastapi-nextauth-jwt/issues/14

To ensure a smoother experience for future users of the library, I have enhanced the README documentation to provide clearer guidance and additional details.

Additionally, I have made some stylistic improvements to make it more visually appealing.

kiwamizamurai committed:
chore; Update README with comprehensive documentation and badges

kiwamizamurai committed:
chore; update readme example code

layzy commented:
👍

kiwamizamurai committed:
docs: Update README with Auth.js v5 cookie name changes and configuration tips

TCatshoek commented on README.md:
```diff
@@ -1,17 +1,33 @@
+
+This project provides a FastAPI dependency for decrypting and validating JWTs generated by Auth.js. It is designed to facilitate the integration of a FastAPI backend with Next.js and NextAuth/Auth.js on the frontend.
+
+> [!WARNING]
```
I think this could be !NOTE instead, as it is not really a warning imo

TCatshoek replied:
I would probably also word it as something like: Using Auth.js with frameworks other than Next.js may work but has not been tested

TCatshoek commented on README.md:
```diff
@@ -1,17 +1,33 @@
+> [!WARNING]
+> It may also be compatible with other frameworks using Auth.js, though this has not been extensively tested.
+
+> [!IMPORTANT]
+> NextAuth.js has been renamed to Auth.js, and the default cookie names have changed to reflect this (e.g., from `next-auth.csrf-token` to `authjs.csrf-token`). You'll need to update your cookie name configuration accordingly if you're using the latest version of Auth.js.
```
This is incorrect, starting from version 2.0.0 the default cookie names take this change into account, see: https://github.com/TCatshoek/fastapi-nextauth-jwt/blob/bef787beff050a4c62a96b721bfbc4d8793d3f33/src/fastapi_nextauth_jwt/fastapi_nextauth_jwt.py#L88

TCatshoek commented on README.md:
```diff
@@ -1,17 +1,33 @@
-in the frontend. It might work with other frameworks using Auth.js too! (But I haven't tested that, feel free to contribute if you do)
+- **JWT Decryption & Validation**: Seamlessly decrypt and validate JWTs generated by Auth.js
+- **CSRF Protection**: Built-in Auth.js-compatible CSRF protection with configurable HTTP methods
+- **Secure by Default**: Automatic secure cookie configuration based on your environment
```
I would prefer to not make any security claims as none of this code has been professionally audited

TCatshoek commented on README.md:
```diff
@@ -21,30 +37,119 @@ from fastapi_nextauth_jwt import NextAuthJWT
 
 JWT = NextAuthJWT(
-    secret="y0uR_SuP3r_s3cr37_$3cr3t",
+    secret="y0uR_SuP3r_s3cr37_$3cr3t",  # Better to use environment variable
```
```suggestion
    secret="y0uR_SuP3r_s3cr37_$3cr3t",  # Leave this out to automatically read the NEXTAUTH_SECRET env var
```

TCatshoek commented on README.md:
```diff
@@ -21,30 +37,119 @@ from fastapi_nextauth_jwt import NextAuthJWT
+async def return_jwt_get(jwt: Annotated[dict, Depends(JWT)]):
+    return {"message": f"Hi {jwt['name']}. Greetings from fastapi!"}
+
+# For CSRF protection testing
+@app.post("/")
+async def return_jwt_post(jwt: Annotated[dict, Depends(JWT)]):
+    return {"message": f"Hi {jwt['name']}. Greetings from fastapi!"}
```
Adding this to the example doesn't really add anything imo

TCatshoek commented on README.md:
```diff
@@ -21,30 +37,119 @@ from fastapi_nextauth_jwt import NextAuthJWT
+
+### Security Options
+
+- **csrf_prevention_enabled** (bool): Toggle CSRF protection
+  - Defaults to `True` in development (`ENV=dev`), `False` otherwise
```
Sorry, this is incorrect in the docstring. It should be: 

```suggestion
- **csrf_prevention_enabled** (bool): Enable CSRF protection
  - Defaults to `False` in development (`ENV=dev`), `True` otherwise
```

TCatshoek commented on README.md:
```diff
@@ -21,30 +37,119 @@ from fastapi_nextauth_jwt import NextAuthJWT
+- **secure_cookie** (bool): Enable secure cookie attributes
+  - Default: `True` (when `NEXTAUTH_URL` starts with https)
+
+- **cookie_name** (str): Session token cookie name
+  - Default: `"__Secure-next-auth.session-token"` (when secure_cookie is True)
+  - Default: `"next-auth.session-token"` (when secure_cookie is False)
+  - For Auth.js (latest), use: `"__Secure-authjs.session-token"` (when secure_cookie is True)
+
+- **csrf_cookie_name** (str): CSRF token cookie name
+  - Default: `"__Host-next-auth.csrf-token"` (when secure_cookie is True)
+  - Default: `"next-auth.csrf-token"` (when secure_cookie is False)
+  - For Auth.js (latest), use: `"__Host-authjs.csrf-token"` (when secure_cookie is True)
```
These already use auth.js names in the latest version

TCatshoek replied:
Please check this again, the cookie names are already the correct authjs ones by default

TCatshoek commented on README.md:
````diff
@@ -21,30 +37,119 @@ from fastapi_nextauth_jwt import NextAuthJWT
+  - Default: `"next-auth.csrf-token"` (when secure_cookie is False)
+  - For Auth.js (latest), use: `"__Host-authjs.csrf-token"` (when secure_cookie is True)
+
+> [!TIP]
+> If you're using the latest version of Auth.js, here's the recommended configuration:
+> ```python
+> JWT = NextAuthJWT(
+>     secret=os.environ["AUTHJS_SECRET"],
+>     cookie_name="__Secure-authjs.session-token",
+>     csrf_cookie_name="__Host-authjs.csrf-token",
+> )
+> ```
````
The code still needs to be adapted to use AUTHJS_SECRET by default, but the cookie names should already be correct by default

TCatshoek commented on README.md:
```diff
@@ -21,30 +37,119 @@ from fastapi_nextauth_jwt import NextAuthJWT
+- `NEXTAUTH_URL`: The URL of your application (affects secure cookie settings)
+- `ENV`: Set to `"dev"` to disable CSRF protection in development
+
+## Contributing
+
+Contributions are welcome! Please feel free to submit a Pull Request.
+
```
I'm very short on time lately, so I can't promise I have time to review PRs in a timely manner :(

TCatshoek commented on README.md:
```diff
@@ -21,30 +37,119 @@ from fastapi_nextauth_jwt import NextAuthJWT
+Contributions are welcome! Please feel free to submit a Pull Request.
+
+
+
+[![Star History Chart](https://api.star-history.com/svg?repos=TCatshoek/fastapi-nextauth-jwt&type=Date)](https://star-history.com/#TCatshoek/fastapi-nextauth-jwt&Date)
```
I'd personally leave this out as it is not really relevant

kiwamizamurai committed:
chore; use !NOTE

kiwamizamurai committed:
chore; modify NOTE sentense

kiwamizamurai committed:
chore; modify IMPORTANT sentence

kiwamizamurai committed:
chore; remove security sentence

kiwamizamurai committed:
chore; modify code comment

kiwamizamurai committed:
chore; remove white space

kiwamizamurai committed:
chore; fix security option documentation

kiwamizamurai committed:
chore; remove documentation

kiwamizamurai committed:
chore; modify TIP code

kiwamizamurai committed:
chore; remove contributing section

kiwamizamurai committed:
chore; remove starhistory

kiwamizamurai committed:
chore; modify example code

kiwamizamurai committed:
chore; fix documentation

kiwamizamurai commented:
@TCatshoek 
Thank you very much for taking the time to review the PR despite your busy schedule. I have addressed all of your comments, and I would appreciate it if you could review the changes again at your convenience.

TCatshoek commented on README.md:
```diff
@@ -1,17 +1,32 @@
+> Using Auth.js with frameworks other than Next.js may work but has not been tested
+
+> [!IMPORTANT]
+> From version 2.0.0, and the default cookie names have changed (like authjs.session-token). You'll need to update your cookie name configuration accordingly if you're using the latest version of Auth.js.
```
Sorry if that was unclear, but this part can be removed entirely. The default cookie name is correct in 2.0.0, and nothing needs to be configured manually

TCatshoek commented on README.md:
```diff
@@ -21,30 +36,102 @@ from fastapi_nextauth_jwt import NextAuthJWT
+- Uses `"A256GCM"` encryption algorithm
+- Sets specific key derivation parameters
+- Uses v4-compatible cookie names
+- Disables salt auto-appending
```
I feel like these details may be a bit too much for someone just using the library, they may not be interested in the inner workings and it could detract from the information they need to know

```suggestion
This provides compatibility with the v4 token format and default settings
```

TCatshoek commented on README.md:
````diff
@@ -21,30 +36,102 @@ from fastapi_nextauth_jwt import NextAuthJWT
+
+### Essential Settings
+
+- **secret** (str): The secret key used for JWT operations. Should match `NEXTAUTH_SECRET` in your Next.js app.
+  ```python
+  JWT = NextAuthJWT(secret=os.getenv("NEXTAUTH_SECRET"))
+  ```
````
Since this is read from NEXTAUTH_SECRET by default, maybe this is more clear
```suggestion
- **secret** (str): The secret key used for JWT operations. Should match `NEXTAUTH_SECRET` in your Next.js app. Leave this out to automatically read the `NEXTAUTH_SECRET` environment variable.
  ```python
  JWT = NextAuthJWT(secret=os.getenv("YOUR_SECRET_ENV_VAR_NAME")))
  ```
```

TCatshoek commented on README.md:
```diff
@@ -21,30 +36,102 @@ from fastapi_nextauth_jwt import NextAuthJWT
+- Basic JWT validation setup
+- CSRF protection configuration
+
+You can also place both the backend and frontend behind a reverse proxy like nginx, as long as cookies reach FastAPI.
```
```suggestion
You can also place both the backend and frontend behind a reverse proxy like nginx, as long as the auth.js cookies reach FastAPI.
```

TCatshoek commented:
Thanks for the work so far! I left some more comments

kiwamizamurai committed:
chore; update readme

kiwamizamurai commented:
@TCatshoek Thank you for your comments! I modified https://github.com/TCatshoek/fastapi-nextauth-jwt/pull/17/commits/23258f128b2a750a54317d06a36965e6f8141336

TCatshoek commented on README.md:
```diff
@@ -21,30 +33,98 @@ from fastapi_nextauth_jwt import NextAuthJWT
+
+This provides compatibility with the v4 token format and default settings
+
+## Security Best Practices
```
```suggestion
## Best Practices
```

TCatshoek commented on README.md:
````diff
@@ -21,30 +33,98 @@ from fastapi_nextauth_jwt import NextAuthJWT
+  JWT = NextAuthJWT(secret=os.getenv("YOUR_SECRET_ENV_VAR_NAME")))
+  ```
+
+### Security Options
+
+- **csrf_prevention_enabled** (bool): Enable CSRF protection
+  - Defaults to `False` in development (`ENV=dev`), `True` otherwise
+
+- **csrf_methods** (Set[str]): HTTP methods requiring CSRF protection
+  - Default: `{'POST', 'PUT', 'PATCH', 'DELETE'}`
+
+### Cookie Configuration
+
+- **secure_cookie** (bool): Enable secure cookie attributes
+  - Default: `True` (when `NEXTAUTH_URL` starts with https)
+
+- **cookie_name** (str): Session token cookie name
+  - Default: `"__Secure-authjs.session-token"` (when secure_cookie is True)
+  - Default: `"authjs.session-token"` (when secure_cookie is False)
+
+- **csrf_cookie_name** (str): CSRF token cookie name
+  - Default: `"__Host-authjs.csrf-token"` (when secure_cookie is True)
+  - Default: `"authjs.csrf-token"` (when secure_cookie is False)
+
+> [!TIP]
+> If you're using the latest version of Auth.js, here's the recommended configuration:
+> ```python
+> JWT = NextAuthJWT(
+>     secret=os.environ["AUTHJS_SECRET"],
+> )
+> ```
+
+### Advanced Options
````
Maybe it would be nice to make it clear that you won't need to touch these settings if you use the default Auth.js settings:
```suggestion
### Additional Options
If your auth.js settings are left at their defaults, you shouldn't need to touch these.

#### Security Options

- **csrf_prevention_enabled** (bool): Enable CSRF protection
  - Defaults to `False` in development (`ENV=dev`), `True` otherwise

- **csrf_methods** (Set[str]): HTTP methods requiring CSRF protection
  - Default: `{'POST', 'PUT', 'PATCH', 'DELETE'}`

#### Cookie Configuration

- **secure_cookie** (bool): Enable secure cookie attributes
  - Default: `True` (when `NEXTAUTH_URL` starts with https)

- **cookie_name** (str): Session token cookie name
  - Default: `"__Secure-authjs.session-token"` (when secure_cookie is True)
  - Default: `"authjs.session-token"` (when secure_cookie is False)

- **csrf_cookie_name** (str): CSRF token cookie name
  - Default: `"__Host-authjs.csrf-token"` (when secure_cookie is True)
  - Default: `"authjs.csrf-token"` (when secure_cookie is False)

> [!TIP]
> If you're using the latest version of Auth.js, here's the recommended configuration:
> ```python
> JWT = NextAuthJWT(
>     secret=os.environ["AUTHJS_SECRET"],
> )
> ```

#### Advanced Options
```

TCatshoek reviewed (changes requested):
Hey, sorry for the delay! Just some small changes and I think we can merge

kiwamizamurai committed:
Update README.md

Co-authored-by: Tom Catshoek <tomcatshoek@zeelandnet.nl>

kiwamizamurai committed:
Update README.md

Co-authored-by: Tom Catshoek <tomcatshoek@zeelandnet.nl>
